third_party/amd/lib/TritonAMDGPUTransforms/RescheduleOps.cpp

```diff
@@ -31,192 +61,404 @@ Operation *createSchedBarrier(OpBuilder &rewriter, Location loc,
+}
+bool opCategoryMem(SchedDagNode *node) {
+  return opCategoryLoad(node) || opCategoryStore(node);
+}
 
-private:
-  void createNodes(Block *mlirBlock) {
+bool opCategoryGlobalLoad(SchedDagNode *node) {
+  Operation *op = node->getOp();
+  return llvm::isa<triton::LoadOp, triton::amdgpu::BufferLoadOp>(op);
+}
+bool opCategoryGlobalStore(SchedDagNode *node) {
+  Operation *op = node->getOp();
+  return llvm::isa<triton::StoreOp>(op);
+}
+
+bool opCategoryNop(SchedDagNode *node) {
+  Operation *op = node->getOp();
+  return llvm::isa<triton::gpu::MemDescSubviewOp, triton::gpu::MemDescTransOp,
+                   tt::amdgpu::ExtractSliceOp, tt::amdgpu::ConcatOp>(op);
+}
+
+bool opCategoryBarrier(SchedDagNode *node) {
+  Operation *op = node->getOp();
+  return llvm::isa<mlir::gpu::BarrierOp, ROCDL::SchedBarrier, ROCDL::SetPrioOp>(
+      op);
+}
+
+/******************************************************************************
+  Creates list of nodes using shared_ptr to keep a single copy.
+  Also creates a list of plain node pointers for algorithm.
+******************************************************************************/
+struct BasicBlockNodeMap {
+  BasicBlockNodeMap(Block *mlirBlock) {
     for (auto it = mlirBlock->begin(); it != mlirBlock->end(); ++it) {
       Operation *op = &(*it);
-      std::unique_ptr<Node> node = std::make_unique<Node>(op);
+      // same as calling new
+      std::shared_ptr<SchedDagNode> node = std::make_shared<SchedDagNode>(op);
+
+      // std::string dbgStr;
+      // llvm::raw_string_ostream dbgStream(dbgStr);
+      // op->print(dbgStream);
+      // LDBG(*node << " -> " << dbgStream.str());
+
       lookup.insert({op, node.get()});
       nodes.push_back(std::move(node));
     }
   }
 
-  enum class Traversal { Topdown, Bottomup };
-  template <Traversal Direction> void insertGPUBarrierEdges() {
+  // Returns a copy of original node list.
+  SchedDagNodeList getNodeList() const {
+    SchedDagNodeList nodeList;
+    for (auto node : nodes) {
+      nodeList.push_back(node.get());
+    }
+    return nodeList;
+  }
+
+  // Lookup node by op.
+  SchedDagNode *operator[](Operation *op) const {
+    if (!lookup.contains(op))
+      return nullptr;
+    return lookup.find(op)->second;
+  }
 
-    auto fwIt = nodes.begin();
-    auto bkIt = nodes.rbegin();
-    auto next = [&]() -> Node * {
-      if constexpr (Direction == Traversal::Topdown) {
-        if (fwIt == nodes.end())
+  llvm::SmallVector<std::shared_ptr<SchedDagNode>> nodes;
+  llvm::MapVector<Operation *, SchedDagNode *> lookup;
+};
+
+/******************************************************************************
+ Dependency is from src/child to dst/parent.
+ dst must preceed src during scheduling.
+ dst / parent
+  ^
+  |
+ src / child
+******************************************************************************/
+struct SchedDep {
+  SchedDagNode *parent; // dst
+  SchedDagNode *child;  // src
```

Comment:
Remove trailing comments. Put it to the docu above

Reply:
Removed.